### ISSUE TITLE
Redstone Writer doesn't trigger block update

### ISSUE THREAD

**Issue by Nagshell** [CyclopsMC/IntegratedDynamics#356]
Would it be possible for Redstone Writer to trigger block update when it changes it's state of output?

In my system I'm using Thermal Dynamics servos to pump liquid, and ID system to control it. My logic is simple enough, and when checking wirter it shows it outputs redstone high, but it doesn't trigger servos.

Triggering block update near servo does the trick, simple act of placing or breaking cobble.
I'll test if having something in between writer and servo will make it work, like repeater, but
it would be awesome if I wouldn't have to do anything.

**Comment by rubensworks** [comment]
It should already do that. What versions are you using?

**Comment by Nagshell** [comment]
1.12.2-0.9.5 of ID and 1.12-2.3.6.13 of thermal, looks like the newest ones.


Attaching screenshot of my current setup - without repeaters signals in redstone writers was chaging, but servos did not respond to that change. After some time even repeaters stopped working - on screenshot both are off, and one writer has simple negation of boolean in the other one.

![2017-11-10_10 57 57](https://user-images.githubusercontent.com/17276063/32652923-4797f200-c606-11e7-9bd2-483ed5ed6ca6.png)

**Comment by rubensworks** [comment]
So are repeaters properly activated by redstone writers or not? I don't fully understand your description.

**Comment by Nagshell** [comment]
No, they aren't properly activated in this configuration.

Removing left writer let the other one work properly.

I also checked chunk boundaries, and writers are in different chunks.

I'm sorry if I'm not clear with my descriptions, English is not my main language.

If you need me to provide more information then tell me what you need to know.

**Comment by rubensworks** [comment]
I can't reproduce this myself at all. So something else must be causing the problem you are describing. I suggest removing mods until you've found the culprit.

**Comment by TheRegulator** [comment]
I Second this bug.
Using Version 1.12.2-0.9.6
on Forge 14.23.0.2531
on Direwolf 20 pack 1.1.0

The Redstonewriter gives an Output but doesn't update the Block in front of it.

As soon, as you update the Block (by placing a block next to it) i get the signal

https://imgur.com/a/tJmwH


<blockquote><img src="https://i.imgur.com/b6CH5I0.jpg?fb" width="48" align="right"><div><img src="//s.imgur.com/images/favicon-16x16.png" height="14"> Imgur</div><div><strong><a href="https://imgur.com/a/tJmwH">Imgur: The most awesome images on the Internet</a></strong></div><div>Imgur: The most awesome images on the Internet.</div></blockquote>

**Comment by rubensworks** [comment]
@TheRegulator Can you try reproducing this without any other mods installed?

**Comment by TheRegulator** [comment]
@rubensworks i tried it:
I could not reproduce it with only Integrate Dynamics and it's dependencies installed,
so i tried it again with the direwolf20 pack in Singleplayer.
There it worked fine too.

Looks like it's only on the Server i play on

**Comment by TvL2386** [comment]
I have the same issue. Playing Direwolf20 1.1.0 (MC1.12.2) on a server hosted by a friend.

Can't reproduce in Single Player mode. Can't even reproduce in Multi Player anymore...
I had an Inventory Reader reading if it was full. I put the variable in the Variable Store.
I created a new Variable Card with the Portable Logic Programmer. It did the Boolean NOT operation on the previous variable. Result: As long as the inventory is NOT full, we emit a redstone signal (to the plant gathererer from Industrial Foregoing).

I am 100% sure I was able to reproduce, which I can't anymore.
What I did to reproduce: lay down some redstone next to a logic cable, place the Redstone Writer on the cable facing the redstone and insert the variable card. The Redstone Writer gui would say: true. So it should emit a redstone signal, however the redstone was emitting lvl 0.
If I would then break the redstone adjacent to the Redstone Writer and place it back, it would be lvl 15.

I had the same thing with my Plant Gatherer. It would not work until I broke the machine and place it back. Same thing with the Plant Sower. Basically this is just the normal MFR Planter/Harvester setup.

Anyways: It seems to work now, and I can't reproduce anymore... Maybe because the server was reloaded in the mean time... Just wanted to inform you guys I've seen it too.

**Comment by TvL2386** [comment]
Oh yeah, one thing: If I would break a block of dirt next to this setup, the redstone would also jump to lvl 15. So I was assuming something with block updates is not going well

**Comment by rubensworks** [comment]
New information by @Gravlok from #436:
____
#### Short description:

Changes of boolean value in Redstone Writer doesn't trigger block update.

#### Expected behaviour:

After inserting empty variable card into Redstone Writer which is connected to path of redstone dust, it should light it up with power of 15.

#### Actual behaviour:

After inserting an empty variable card into Redstone Writer nothing happens with redstone dust unless a manual block update is performed. Same thing occurs when trying to turn it off. Taking out empty card, doesn't make the redstone signal to go down unless a manual block update is performed.

#### Steps to reproduce the problem:

1. Place a piece of redstone dust
2. Place Redstone Writer pointing at redstone dust, add a piece of cable to it
3. Insert empty variable card or boolean true variable into Redstone Writer
4. Place anything near the redstone to force update block
5. To confirm it's not visual bug, place Redstone Reader attached to redstone dust, take it's redstone integer variable, and insert in Display Panel to check power level

#### Additional notes
After further examination of the problem I narrowed down the possible causes. I play on DW20 1.12 modpack which have disabled energy usage by default (in config I:general.energyConsumptionMultiplier=0). In the reproduce steps I purposely skipped "add the battery and connect it to energy source". When the network is powered regardless of the config, the bug above doesn't happen. HOWEVER! If the Redstone Writer is placed BEFORE the network was plugged into power source the bug still happens. What's interesting, inserting an integer card with any value in it works fine, regardless if powered or not.
____
#### Versions:

- This mod: 1.12.2-0.9.6
- CyclopsCore: 1.12.2-0.10.23
- Minecraft: .1.12.2
- Forge: 14.23.1.2555

#### Log file:

N/A

**Comment by rubensworks** [comment]
This new information doesn't help me reproducing this unfortunately. It works just fine for in singleplayer and multiplayer. I suggest to report this to the modpack developer, as there must be another mod in the pack that somehow messes with how redstone updates are propagated.

**Comment by Suterusu1337** [comment]
I have a world where this is both working and not with the same block update issue.  I can provide a dropbox of the world if you would like to review.

I am still working to reproduce the problem.  It affects both vanilla redstone and modded blocks.
Versions:

- **This mod: 1.12.2-0.10.0**
- CyclopsCore: 1.12.2-0.11.0
- Minecraft: 1.12.2
- Forge: 14.23.1.2555

**Comment by rubensworks** [comment]
@Suterusu1337 Is it a (minimal) world with just ID and dependencies installed? If so, yes please :-)

**Comment by Suterusu1337** [comment]
@rubensworks modded, I can upload a full playable zip, or a video of the issue

**Comment by rubensworks** [comment]
@Suterusu1337 A video would be nice. Could you then also send the server and client logs during the video?

**Comment by Suterusu1337** [comment]
Working https://youtu.be/Ir4fDlwfeSE

Not working https://youtu.be/20c04D2-atA

Not working 2 https://youtu.be/0VqtNLn1FAw

I cannot upload fml-client-latest.log to pastebin due to file size (2,412KB), what means should I deliver it?

latest.log
https://pastebin.com/vT4iU0Zr



<blockquote><img src="https://i.ytimg.com/vi/Ir4fDlwfeSE/maxresdefault.jpg" width="48" align="right"><div><img src="/yts/img/favicon_144-vfliLAfaB.png" height="14"> YouTube</div><div><strong><a href="https://www.youtube.com/watch?v=Ir4fDlwfeSE">RS Writer Working</a></strong></div><div> </div></blockquote>
<blockquote><img src="https://i.ytimg.com/vi/20c04D2-atA/maxresdefault.jpg" width="48" align="right"><div><img src="/yts/img/favicon_144-vfliLAfaB.png" height="14"> YouTube</div><div><strong><a href="https://www.youtube.com/watch?v=20c04D2-atA">RS Writer Non-Working 1</a></strong></div><div> </div></blockquote>
<blockquote><img src="https://i.ytimg.com/vi/0VqtNLn1FAw/maxresdefault.jpg" width="48" align="right"><div><img src="/yts/img/favicon_144-vfliLAfaB.png" height="14"> YouTube</div><div><strong><a href="https://www.youtube.com/watch?v=0VqtNLn1FAw">RS Writer Non-Working 2</a></strong></div><div> </div></blockquote>
<blockquote><img src="https://pastebin.com/i/facebook.png" width="48" align="right"><div>Pastebin</div><div><strong><a href="https://pastebin.com/vT4iU0Zr">latest.log - Pastebin.com</a></strong></div></blockquote>

**Comment by Suterusu1337** [comment]
I can set more of the issue not working that could be more clear if needed.

**Comment by rubensworks** [comment]
@Suterusu1337 Thanks for the videos.

It seems that the issue always occurs on AA oil generators. Is this always the case? Or have you encountered situations where the oil generators do properly respond to redstone changes?

~~It's possible that these machines (and other machines) require a block update before re-reading the redstone values, while this should not be required (a simple neighbour change event should be sufficient). What I could do to resolve this, is add an option to the redstone writer to trigger an actual block update on redstone changes, even though this would lead to more server+client lag.~~

EDIT: Just tried this myself with these generators, and it works just fine for me. So the issue must lie somewhere else.

**Comment by rubensworks** [comment]
@Suterusu1337 Would you be willing to help out by trying out the following dev build?
https://www.dropbox.com/s/zwf04tygfifu9er/IntegratedDynamics-1.12.2-0.10.0-DEV.jar?dl=0

It will print some debug information to the logs regarding redstone updates.
If possible, could you make your system trigger some redstone changes, and send me the logs?
That should help me pinpoint the exact point of failure.

<blockquote><img src="https://cfl.dropboxstatic.com/static/images/icons128/page_white.png" width="48" align="right"><div>Dropbox</div><div><strong><a href="https://www.dropbox.com/s/zwf04tygfifu9er/IntegratedDynamics-1.12.2-0.10.0-DEV.jar?dl=0">IntegratedDynamics-1.12.2-0.10.0-DEV.jar</a></strong></div><div>Shared with Dropbox</div></blockquote>

**Comment by Suterusu1337** [comment]
@rubensworks which log file(s) would you like?

**Comment by rubensworks** [comment]
@Suterusu1337 latest.log should be fine. You should see a lot of debug output in there.

**Comment by Suterusu1337** [comment]
I cannot load this into an existing world, perhaps the CyclopsCore lib needs to be your DEV version as well?

https://pastebin.com/VmesZN2y


<blockquote><img src="https://pastebin.com/i/facebook.png" width="48" align="right"><div>Pastebin</div><div><strong><a href="https://pastebin.com/VmesZN2y">Crash -1 - Pastebin.com</a></strong></div></blockquote>

**Comment by rubensworks** [comment]
@Suterusu1337 Ah, sorry, you'll need a new version of IT as well: https://oss.jfrog.org/artifactory/simple/libs-release/org/cyclops/integratedtunnels/IntegratedTunnels/1.12.2-1.4.0-229/IntegratedTunnels-1.12.2-1.4.0-229.jar

Make sure to backup your world though!

**Comment by Suterusu1337** [comment]
Using the DEV files, there are no DEBUG outputs in any log.  I am running this in SinglePlayer mode if that makes a difference.  Is there a config option I need to enable?

I replicated both working and non-working conditions, neither produce any debug information.

**Comment by Suterusu1337** [comment]
@rubensworks 
Here is the 210MB log file as a 4MB zip:
[log.zip](https://github.com/CyclopsMC/IntegratedDynamics/files/1608637/log.zip)

I think these are the correct positions:

**Not working setup:**
Oil Generator	        x-215	z266	y86
Redstone Writer	x-216	z266	y86

**Working setup:**
Redstone Writer	x-204	z260	y87
Wire		                x-205	x260	y87

Interestingly, with the DEV version, I have to break + replace some Immersive Engineering cables to get the RF transfer to work again, perhaps a coincidence, unsure.

**Comment by rubensworks** [comment]
@Suterusu1337 Thanks for the logs!

Due to there being several writers that change values frequently, it's impossible to track down the not-working-part when it changes values.
I just changed some of the debug code to make this easier to identify.

Could you send me the logs again with this build?: https://www.dropbox.com/s/kosamq1nq47p2sy/IntegratedDynamics-1.12.2-0.10.0-DEV.jar?dl=0

<blockquote><img src="https://cfl.dropboxstatic.com/static/images/icons128/page_white.png" width="48" align="right"><div>Dropbox</div><div><strong><a href="https://www.dropbox.com/s/kosamq1nq47p2sy/IntegratedDynamics-1.12.2-0.10.0-DEV.jar?dl=0">IntegratedDynamics-1.12.2-0.10.0-DEV.jar</a></strong></div><div>Shared with Dropbox</div></blockquote>

**Comment by Suterusu1337** [comment]
@rubensworks 
[log2.zip](https://github.com/CyclopsMC/IntegratedDynamics/files/1613675/log2.zip)

**Comment by rubensworks** [comment]
@Suterusu1337 Thanks!
According to my debug output, all block updates are properly being called, so the problem must lie somewhere else.

I just made a small change that may be related to it, could you try if this build (no debug output anymore) fixes the problem for you?: https://www.dropbox.com/s/2gs53cjsp22y7q4/IntegratedDynamics-1.12.2-0.10.0-DEV.jar?dl=0

<blockquote><img src="https://cfl.dropboxstatic.com/static/images/icons128/page_white.png" width="48" align="right"><div>Dropbox</div><div><strong><a href="https://www.dropbox.com/s/2gs53cjsp22y7q4/IntegratedDynamics-1.12.2-0.10.0-DEV.jar?dl=0">IntegratedDynamics-1.12.2-0.10.0-DEV.jar</a></strong></div><div>Shared with Dropbox</div></blockquote>

**Comment by ghost** [comment]
It worked for a minute with the dev version, but then the bug happened again and also I couldn't open vanilla chests anymore

~~Edit: Seems like the chests not opening thing was because of foamfix according to my latest log, maybe that's the mod causing all this?*~~

Apparently the chest bug was because I was using a bugged version of forge, anyway the dev version didn't fix it for me.

**Comment by Suterusu1337** [comment]
The DEV version is working for me on the setup with the AE Oil Gen.

**Comment by rubensworks** [comment]
@Suterusu1337 Thanks for testing!

@cowardlyDog Any details you can give me about your setup? Is it not working in some cases? Or all cases? What happens if you just power redstone dust?

**Comment by rubensworks** [comment]
@cowardlyDog Feel free to comment if there's anything else you could tell me. If not, I'm considering this (finally) fixed.

**Comment by ShadowDrakken** [comment]
Confirmed I'm having this same issue with 0.10 in the FTB Horizons 3 pack. Simple setup: writer, single cable, reader. Writer pointed at cobblestone with strong signal (or Cyclic fan with any signal, since it's the block I'm attempting to affect), lever on reader, and a variable card set to the reader and dropped into writer.

Once the signal is turned on, there's no way to turn it back off. The writer itself shows that it's toggling, but it continues to push a signal to the adjacent block.

The caveat here though is that if I put redstone dust next to the reader instead of a block, it works as expected. This has me wondering if it's actually a mod bug, or a Forge bug. If it's a Forge bug then it needs to be reported over there.

**Comment by ShadowDrakken** [comment]
Ok, I found out what's going on. Even though it's updating the redstone signal, it's not actually sending a block update for whatever reason. If I update the block next to either the strong powered block, or next to something weakly powered by the strong powered block then the signal updates as well. It's behaving like an old school BUD.

**Comment by rubensworks** [comment]
@ShadowDrakken You can try the latest dev build, that should have fixed it.

**Comment by ShadowDrakken** [comment]
I'll have to wait till it's pushed to Twitch, it impacts everyone on my server.

Also, above someone said it didn't fix it for them.

**Comment by drcrazy** [comment]
I have integrated dynamics 1.12.2-0.10.1 within DW20 1.12 1.7.1 pack. 

Redstone writer fails to update output (provide signal or stop provide signal).

Setup: writer -> redstone -> Thermal rs relay
![image](https://user-images.githubusercontent.com/1382352/35174637-1ae7a2c0-fd81-11e7-92cb-bde98596af21.png)

writer has a variable which is currently 'true' but still no rs output. 
![image](https://user-images.githubusercontent.com/1382352/35174667-45942598-fd81-11e7-9853-45ba8cf516a2.png)


Changing signal strength from weak to strong, insert\remove variable doesn't help. 

I have also used writers in another automation setup and they were fine. World save is 500Mb. Want me to provide it?
